tests/test_execution.py

```diff
@@ -0,0 +1,213 @@
+
+    @property
+    def loaded_config_sources(self):
+        return []
+
+    @property
+    def run_config(self):
+        return RunConfiguration()
+
+    @property
+    def plugin_cache(self):
+        return MockPluginCache()
+
+    def __init__(self, *args, **kwargs):
+        super(MockConfigManager, self).__init__(*args, **kwargs)
+        self._joblist = None
+        self._run_config = RunConfiguration()
+
+    def to_pod(self):
+        return {}
+
+
+class MockPluginCache(Mock):
+
+    def list_plugins(self, kind=None):
+        return []
+
+
+class MockProcessorManager(Mock):
+
+    def __init__(self, *args, **kwargs):
+        super(MockProcessorManager, self).__init__(*args, **kwargs)
+
+    def get_enabled(self):
+        return []
+
+
+class Job_force_retry(Job):
+    '''This class imitates a job that fails once then passes on retry'''
+    @property
+    def status(self):
+        return self.state.status
+
+    @status.setter
+    def status(self, value):
+        if (self.retries != self.num_retries) and (value == Status.RUNNING):
+            self.state.status = Status.FAILED
+            if self.output:
+                self.output.status = Status.FAILED
+        else:
+            self.state.status = value
+            if self.output:
+                self.output.status = value
+        self.state.timestamp = datetime.utcnow()
+
+    def __init__(self, retries, *args, **kwargs):
+        self.num_retries = retries
+        super(Job_force_retry, self).__init__(*args, **kwargs)
+
+
+class TestRunState(TestCase):
+
+    def setUp(self):
+        self.path = tempfile.mkstemp()[1]
+        os.remove(self.path)
+        self.initialise_signals()
+
+        config = MockConfigManager()
+        output = init_run_output(self.path, config)
+
+        self.context = ExecutionContext(config, Mock(), output)
+
+        self.job_spec = JobSpec()
+        self.job_spec.augmentations = {}
+        self.job_spec.finalize()
+
+    def tearDown(self):
+        signal.disconnect(self._verify_serialized_state, signal.RUN_INITIALIZED)
+        signal.disconnect(self._verify_serialized_state, signal.JOB_STARTED)
+        signal.disconnect(self._verify_serialized_state, signal.JOB_RESTARTED)
+        signal.disconnect(self._verify_serialized_state, signal.JOB_COMPLETED)
+        signal.disconnect(self._verify_serialized_state, signal.JOB_FAILED)
+        signal.disconnect(self._verify_serialized_state, signal.JOB_ABORTED)
+        signal.disconnect(self._verify_serialized_state, signal.RUN_FINALIZED)
+
+    def test_job_state_transitions_pass(self):
+        '''Tests state equality when the job passes first try'''
+        job = Job(self.job_spec, 1, self.context)
+        job.workload = Mock()
+
+        self.context.cm._joblist = [job]
+        self.context.run_state.add_job(job)
+
+        runner = Runner(self.context, MockProcessorManager())
+        runner.run()
+
+    def test_job_state_transitions_fail(self):
```

Comment:
Now this test seems to be failing instead.
(https://travis-ci.org/github/ARM-software/workload-automation/jobs/705788405)